src/main/java/com/aws/greengrass/mqtt/bridge/clients/LocalMqtt5Client.java

```diff
@@ -0,0 +1,476 @@
+        synchronized (subscriptionsLock) {
+            toSubscribeLocalMqttTopics.clear();
+            toSubscribeLocalMqttTopics.addAll(topics);
+
+            LOGGER.atDebug().kv(LOG_KEY_TOPICS, topics).log("Updated local MQTT5 topics to subscribe");
+            if (client.getIsConnected()) {
+                updateSubscriptionsInternalAsync();
+            }
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void updateSubscriptionsInternalAsync() {
+        synchronized (subscriptionsLock) {
+            cancelUpdateSubscriptionsTask();
+
+            Set<String> topicsToRemove = new HashSet<>(subscribedLocalMqttTopics);
+            topicsToRemove.removeAll(toSubscribeLocalMqttTopics);
+
+            Set<String> topicsToSubscribe = new HashSet<>(toSubscribeLocalMqttTopics);
+            topicsToSubscribe.removeAll(topicsToRemove);
+            topicsToSubscribe.removeAll(subscribedLocalMqttTopics);
+
+            updateSubscriptionsTask = executorService.submit(() -> {
+                for (String topic : topicsToRemove) {
+                    if (Thread.currentThread().isInterrupted()) {
+                        return;
+                    }
+                    unsubscribe(topic);
+                    // TODO retry unsubscribe failures
+                }
+                LOGGER.atDebug().kv(LOG_KEY_TOPICS, topicsToSubscribe).log("Subscribing to MQTT topics");
+                subscribeToTopics(topicsToSubscribe);
+            });
+        }
+    }
+
+    private void cancelUpdateSubscriptionsTask() {
+        synchronized (subscriptionsLock) {
+            if (updateSubscriptionsTask != null) {
+                updateSubscriptionsTask.cancel(true);
+            }
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void subscribeToTopics(Set<String> topics) {
+        for (String topic : topics) {
+            try {
+                RetryUtils.runWithRetry(mqttExceptionRetryConfig, () -> {
+                    subscribe(topic); // TODO retry based on return code
+                    return null;
+                }, "subscribe-mqtt5-topic", LOGGER);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return;
+            } catch (Exception e) {
+                LOGGER.atError().setCause(e).kv(LOG_KEY_TOPIC, topic).log("Failed to subscribe");
+            }
+        }
+    }
+
+    private void subscribe(String topic) {
+        if (!client.getIsConnected()) {
+            return;
+        }
+        SubscribePacket subscribePacket = new SubscribePacket.SubscribePacketBuilder()
+                // TODO other mqtt5-specific fields
+                .withSubscription(topic, QOS.AT_LEAST_ONCE).build();
+        LOGGER.atDebug().kv(LOG_KEY_TOPIC, topic).log("Subscribing to MQTT topic");
+        try {
+            SubAckPacket subAckPacket = client.subscribe(subscribePacket).get();
```

Comment:
What do we do now with paho? I don't think we should wait indefinitely, some timeout should be applied.

Reply:
w/ paho we don't timeout the operation 😬